### DISCUSSION TITLE
Accessing local files

### DISCUSSION THREAD

**Discussion by mariussdl** [pyscript/pyscript#1438] [Q&A] [answered]
Hey guys! 

I'm working with a small database written in JSON, so it's stored locally in a seperate file.
I'm trying to access this file in a `py-script` tag inside a context manager, but I always get an error saying that my file does not exist. Even though the file exists and it's the correct spelling of the name.

If any one can help me, I'd really appreciate it.
Thanks

--therealmarius

**Comment by coder0107git** [comment]
By _locally_, do you mean on your OS (filesystem) or loaded into pyscript? The browser doesn't give websites access to user files without using a special API or a file picker. If the file is stored in pyscript, you would need to check if the file was successfully loaded.

**Edit:** Looks like I read your question wrong. You'll need to either (1) embed the JSON in your HTML (or python) file, or (2) make the file accessible from a webserver, get the file contents, load it into pyscript and then access it from python. That would end up looking something like the following where `<file>` is the web accessible path.
``` javascript
// Based on code from https://stackoverflow.com/a/63966487
const data = new Uint8Array(await (await fetch("<file>")).arrayBuffer());

const filename = "myfile.json";
const stream = FS.open("myfile.json", "w+");
FS.write(stream, data, 0, data.length, 0);
FS.close(stream);
```
For more information about the websites/pyscript/pyodide filesystem limitations, please read [this](https://pyodide.org/en/stable/usage/faq.html#why-can-t-i-load-files-from-the-local-file-system).

**Reply by mariussdl** [reply]
Hey! Thanks for your answer.
For the moment I don't use JavaScript. How can I embed my file into my HTML ? Does it mean that my file will be loaded into PyScript ?

Because when I lunch my script I got this error:
```
Traceback (most recent call last):
  File "/lib/python3.10/_pyodide/_base.py", line 460, in eval_code
    .run(globals, locals)
  File "/lib/python3.10/_pyodide/_base.py", line 306, in run
    coroutine = eval(self.code, globals, locals)
  File "<exec>", line 12, in <module>
FileNotFoundError: [Errno 44] No such file or directory: 'pearltrees.json'
```
So I belive it's about loading the file into PyScript.

**Comment by JeffersGlass** [comment, marked as answer]
There is functionality within the `<py-config>` tag for loading additional files into the in-browser file system, where PyScript can access them. This functionality is used by adding what we call [[[fetch]] configurations](https://docs.pyscript.net/latest/reference/elements/py-config.html#local-modules) to py-config.

For example, with the following local file structure:
```
demo.html
my_local_folder/
├─ my_file.txt
```
Assuming I have a [simple local server](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server) serving this folder, I could load the file into the in-browser filesystem using the following PyScript code:
```html
<head>
    <script defer src="https://pyscript.net/releases/2023.03.1/pyscript.js"></script>
    <link rel="stylesheet" href="https://pyscript.net/releases/2023.03.1/pyscript.css">
</head>
<body>
    <py-config>
        [[fetch]]
        files = ['myfile.txt']
        from = 'my_local_folder'
        to_folder = 'data'
    </py-config>
    <py-script>
        with open("data/myfile.txt", 'r', encoding='utf-8') as f:
            print(f.read())
    </py-script>
</body>
</html>
```

Be aware, however, that this method (and indeed the \<file\> one mentioned elsewhere) fetch the _contents_ that file and create a _copy_ of it in the virtual in-browser file system; you can write changes to that file just like any other text file, but they won't have an effect on the source where that content came from. To 'persist' the changes back to the place the json file came from, you'll either want some kind of web server/API that you can write to, or for a simple case, allow the user to re-download the JSON file.

---

As a side note, there is also the [File System Access API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API), which prompts the user and, if allowed, can provide direct read and/or write access to files on the local disk. However, it has limited availability in browsers at this point, and is only available in secure contexts. I don't personally think it's worth deploying with it at this point, though it's fun to play with.

**Reply by mariussdl** [reply]
That's exactly what I needed! Thanks!
